hostsfile/src/lib.rs

```diff
@@ -218,14 +252,80 @@ impl HostsBuilder {
             writeln!(&mut s, "{}", line)?;
         }
 
+        match Self::write_and_swap(&temp_path, hosts_path, &s) {
+            Err(_) => {
+                Self::write_clobber(hosts_path, &s)?;
+                log::debug!("wrote hosts file with the clobber fallback strategy");
+            },
+            _ => {
+                log::debug!("wrote hosts file with the write-and-swap strategy");
+            },
+        }
+        Ok(())
+    }
+
+    fn write_and_swap(temp_path: &Path, hosts_path: &Path, contents: &[u8]) -> io::Result<()> {
+        std::fs::copy(&hosts_path, &temp_path)?;
+        OpenOptions::new()
+            .create(false)
+            .read(true)
+            .write(true)
+            .truncate(true)
+            .open(&temp_path)?
+            .write_all(contents)?;
+
+        std::fs::rename(temp_path, hosts_path)?;
+        Ok(())
+    }
+
+    fn write_clobber(hosts_path: &Path, contents: &[u8]) -> io::Result<()> {
         OpenOptions::new()
             .create(true)
             .read(true)
             .write(true)
             .truncate(true)
             .open(hosts_path)?
-            .write_all(&s)?;
-
+            .write_all(contents)?;
         Ok(())
     }
```

Comment:
Do we actually need to `std::fs::copy(&hosts_path, &temp_path)` in `write_and_swap()`? If I'm not mistaken we've already _read_ hosts file up in `write_to()`.

We should be able to call `write_clobber(temp_path, contents)` in `write_and_swap()`.

Alternatively, if we want to prevent a collision of temporary file names, we could use different more defensive OpenOptions in the two:
write_and_swap(): `OpenOptions::new().`**`create_new()`**`.write(true)`
write_clobber(): `OpenOptions::new().create(true).write(true).truncate(true)`

Reply:
I added the `std::fs::copy(&hosts_path, &temp_path)` back in since it was in a previous version of `hostsfiles` as a simple way to preserve permissions from the file that it would be overwriting. Seemed reasonable, especially since hostsfiles tend to be small and CoW filesystems are getting more prevalent anyway :).

Good point about using `write_clobber` within `write_and_swap` - done!

Reply:
> I added the `std::fs::copy(&hosts_path, &temp_path)` back in since it was in a previous version of `hostsfiles` as a simple way to preserve permissions from the file that it would be overwriting.

Oh, that's a nice trick! We could optionally add a comment to explain, but that's a super-nit.

Reply:
Will do! You're right, that line definitely deserves an explanation.